pkg/featuregates/featuregate.go

```diff
@@ -0,0 +1,101 @@
+package featuregates
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	configv1 "github.com/openshift/api/config/v1"
+)
+
+// FeatureGate indicates whether a given feature is enabled or not
+// This interface is directly taken from openshift/library-go and modified as required.
+type FeatureGate interface {
+	// Enabled returns true if the key is enabled.
+	Enabled(key configv1.FeatureGateName) bool
+}
+
+// featureGate holds list of enabled and disabled features of a featureSet.
+type featureGate struct {
+	enabled  sets.Set[configv1.FeatureGateName]
+	disabled sets.Set[configv1.FeatureGateName]
+}
+
+// NewFeatureGate initializes the enabled and disabled features list.
+func NewFeatureGate(enabled, disabled []configv1.FeatureGateName) FeatureGate {
+	return &featureGate{
+		enabled:  sets.New[configv1.FeatureGateName](enabled...),
+		disabled: sets.New[configv1.FeatureGateName](disabled...),
+	}
+}
+
+// Enabled returns true when a featureGate is enabled.
+func (f *featureGate) Enabled(key configv1.FeatureGateName) bool { return f.enabled.Has(key) }
+
+// GenerateCustomFeatures generates the custom feature gates from the install config.
+func GenerateCustomFeatures(features []string) (*configv1.CustomFeatureGates, error) {
+	customFeatures := new(configv1.CustomFeatureGates)
+	knownFeatures := getKnownFeatures()
+
+	var errs []error
+	for _, feature := range features {
+		featureName, enabled, err := parseCustomFeatureGate(knownFeatures, feature)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		if enabled {
+			customFeatures.Enabled = append(customFeatures.Enabled, featureName)
+		} else {
+			customFeatures.Disabled = append(customFeatures.Disabled, featureName)
+		}
+	}
+	if len(errs) != 0 {
+		return nil, fmt.Errorf("failed to parse custom features: %s", errs)
+	}
+
+	return customFeatures, nil
+}
+
+// getKnownFeatures returns the list of all known FeatureGates.
+func getKnownFeatures() sets.Set[configv1.FeatureGateName] {
+	knownFeatures := sets.New[configv1.FeatureGateName]()
+
+	for _, featureSet := range configv1.FeatureSets {
+		for _, feature := range featureSet.Enabled {
+			knownFeatures.Insert(feature.FeatureGateAttributes.Name)
+		}
+		for _, feature := range featureSet.Enabled {
+			knownFeatures.Insert(feature.FeatureGateAttributes.Name)
+		}
```

Comment:
Are these duplicates on purpose?

Reply:
No, sorry, it's an error from my side, second loop is for Disabled features. I will correct it.